From 235e4867bd00563614f6b7503fe4d92eaf049bf2 Mon Sep 17 00:00:00 2001
From: Amir Vadai <avadai@redhat.com>
Date: Thu, 11 Sep 2014 11:59:52 -0400
Subject: [ethernet] mlx4: Pad ethernet packets smaller than 17 bytes

Message-id: <12334223e82cae98a7af787fcbe0625c5ae53dce.1410425016.git.avadai@redhat.com>
Patchwork-id: 93607
O-Subject: [Patch RHEL7.1 V1 056/197] net/mlx4_en: Pad ethernet packets smaller than 17 bytes
Bugzilla: 1107617 1107618 1107619 1107620
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>

From: Eugenia Emantayev <eugenia@mellanox.com>

Hardware can't accept packets smaller than 17 bytes. Therefore need to
pad with zeros.

Signed-off-by: Amir Vadai <amirv@mellanox.com>
Signed-off-by: Eugenia Emantayev <eugenia@mellanox.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
(cherry picked from commit 93591aaa62f89820f4ae0558f01eaf9a359738da)
Signed-off-by: Amir Vadai <avadai@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/mellanox/mlx4/en_tx.c b/drivers/net/ethernet/mellanox/mlx4/en_tx.c
index 3a3c8fb..eae35a9 100644
--- a/drivers/net/ethernet/mellanox/mlx4/en_tx.c
+++ b/drivers/net/ethernet/mellanox/mlx4/en_tx.c
@@ -585,7 +585,13 @@ static void build_inline_wqe(struct mlx4_en_tx_desc *tx_desc, struct sk_buff *sk
  int spc = MLX4_INLINE_ALIGN - CTRL_SIZE - sizeof *inl;
 
  if (skb->len <= spc) {
-  inl->byte_count = cpu_to_be32(1 << 31 | skb->len);
+  if (likely(skb->len >= MIN_PKT_LEN)) {
+   inl->byte_count = cpu_to_be32(1 << 31 | skb->len);
+  } else {
+   inl->byte_count = cpu_to_be32(1 << 31 | MIN_PKT_LEN);
+   memset(((void *)(inl + 1)) + skb->len, 0,
+          MIN_PKT_LEN - skb->len);
+  }
   skb_copy_from_linear_data(skb, inl + 1, skb_headlen(skb));
   if (skb_shinfo(skb)->nr_frags)
    memcpy(((void *)(inl + 1)) + skb_headlen(skb), fragptr,
-- 
1.7.1